### DISCUSSION TITLE
Failed to Install the DeePMD-kit’s C++ interface

### DISCUSSION THREAD

**Discussion by qdgp** [deepmodeling/deepmd-kit#428] [Q&A]
I am trying to install the DeePMD-kit’s C++ interface, and after execute 'make', it gives following error. 
The deepmd-kit source code is 1.3.2, gcc is gcc_linux-64=7.5.0 from conda.

> [  2%] Building NVCC (Device) object op/cuda/CMakeFiles/deepmd_op_cuda.dir/deepmd_op_cuda_generated_prod_virial_se_r.cu.o
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(177): error: "conjunction_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(177): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(180): error: "disjunction_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(180): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(183): error: "negation_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(183): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2661): error: "is_swappable_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2661): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2666): error: "is_nothrow_swappable_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2666): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2745): error: "is_swappable_with_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2745): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2750): error: "is_nothrow_swappable_with_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2750): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2882): error: "is_invocable_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2882): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2886): error: "is_nothrow_invocable_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2886): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2891): error: "is_invocable_r_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2891): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2896): error: "is_nothrow_invocable_r_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2896): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2903): error: "is_void_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2903): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2905): error: "is_null_pointer_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2905): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2907): error: "is_integral_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2907): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2909): error: "is_floating_point_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2909): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2911): error: "is_array_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2911): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2913): error: "is_pointer_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2913): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2915): error: "is_lvalue_reference_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2915): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2918): error: "is_rvalue_reference_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2918): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2921): error: "is_member_object_pointer_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2921): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2924): error: "is_member_function_pointer_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2924): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2927): error: "is_enum_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2927): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2929): error: "is_union_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2929): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2931): error: "is_class_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2931): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2933): error: "is_function_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2933): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2935): error: "is_reference_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2935): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2937): error: "is_arithmetic_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2937): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2939): error: "is_fundamental_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2939): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2941): error: "is_object_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2941): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2943): error: "is_scalar_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2943): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2945): error: "is_compound_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2945): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2947): error: "is_member_pointer_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2947): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2949): error: "is_const_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2949): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2951): error: "is_volatile_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2951): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2953): error: "is_trivial_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2953): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2955): error: "is_trivially_copyable_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2955): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2958): error: "is_standard_layout_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2958): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2960): error: "is_pod_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2960): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2962): error: "is_literal_type_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2962): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2964): error: "is_empty_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2964): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2966): error: "is_polymorphic_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2966): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2968): error: "is_abstract_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2968): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2970): error: "is_final_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2970): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2972): error: "is_signed_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2972): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2974): error: "is_unsigned_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2974): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2976): error: "is_constructible_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2976): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2979): error: "is_default_constructible_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2979): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2982): error: "is_copy_constructible_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2982): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2985): error: "is_move_constructible_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2985): error: "constexpr" is not valid here
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2988): error: "is_assignable_v" is not a function or static data member
> 
> /mnt/ufs18/home-183/gepei/anaconda3/envs/tensorflow_GPU/x86_64-conda-linux-gnu/include/c++/7.5.0/type_traits(2988): error: "constexpr" is not valid here
> 
> Error limit reached.
> 100 errors detected in the compilation of "/tmp/tmpxft_00006d67_00000000-15_prod_virial_se_r.compute_75.cpp1.ii".
> Compilation terminated.
> CMake Error at deepmd_op_cuda_generated_prod_virial_se_r.cu.o.release.cmake:282 (message):
>   Error generating file
>   /mnt/home/gepei/software/deepmd_GPU/deepmd-kit/source/build/op/cuda/CMakeFiles/deepmd_op_cuda.dir//./deepmd_op_cuda_generated_prod_virial_se_r.cu.o
> 
> 
> make[2]: *** [op/cuda/CMakeFiles/deepmd_op_cuda.dir/deepmd_op_cuda_generated_prod_virial_se_r.cu.o] Error 1
> make[1]: *** [op/cuda/CMakeFiles/deepmd_op_cuda.dir/all] Error 2
> make: *** [all] Error 2

Could you help me with that?
Thank you.

**Comment by felix5572** [comment]
Could you provide the Nvcc version  cuda version. And check the  c++ standard used (c++11 c++14 c++17 ) while compiling?
It looks like the mismatch of c++ standard. And some syntax in the file 'type_traits' cannot be correctly recognized by the compiler.

**Reply by qdgp** [reply]
I have solved this by using gcc 8.3.0.

**Comment by njzjz** [comment]
You need to change your `CXXFLAGS` from `-std=c++17` to `-std=c++11`. See #755.
